velox/docs/develop/debugging/print-expr-with-stats.rst

```diff
@@ -0,0 +1,148 @@
+individual expressions and introduced a helper function printExprWithStats to
+print the expression tree annotated with runtime stats. This allows us to
+identify bottlenecks in the individual functions.
+
+printExprWithStats
+------------------
+
+This is a helper method that can be used to print out the expression tree
+annotated with runtime statistics for individual expressions. The stats include
+CPU time and number of processed rows.
+
+For example, here is what the output of printExprWithStats may look like for an
+expression (c0 + 3) * c1 evaluated on 1024 rows represented using flat
+vectors.
+
+.. code-block::
+
+    multiply [cpu time: 53.70us, rows: 1024] -> BIGINT
+       plus [cpu time: 70.69us, rows: 1024] -> BIGINT
+          cast(c0 as BIGINT) [cpu time: 132.21us, rows: 1024] -> BIGINT
+             c0 [cpu time: 0ns, rows: 0] -> INTEGER
+          3:BIGINT [cpu time: 0ns, rows: 0] -> BIGINT
+       cast(c1 as BIGINT) [cpu time: 67.17us, rows: 1024] -> BIGINT
+          c1 [cpu time: 0ns, rows: 0] -> INTEGER
+
+If the same expression was evaluated on dictionary-encoded data, the stats would
+show that evaluation processed fewer rows (i.e. number of rows with unique
+values) and used less CPU time:
+
+.. code-block::
+
+    multiply [cpu time: 35.52us, rows: 205] -> BIGINT
+       plus [cpu time: 46.66us, rows: 205] -> BIGINT
+          cast(c0 as BIGINT) [cpu time: 80.15us, rows: 205] -> BIGINT
+             c0 [cpu time: 0ns, rows: 0] -> INTEGER
+          3:BIGINT [cpu time: 0ns, rows: 0] -> BIGINT
+       cast(c1 as BIGINT) [cpu time: 34.44us, rows: 205] -> BIGINT
+          c1 [cpu time: 0ns, rows: 0] -> INTEGER
+
+printExprWithStats can be used to explore various optimizations in the
+expression evaluation and investigate whether these applied in a particular
+run.
+
+For example, we can demonstrate common sub-expression elimination using an
+expression set with two expressions: (c0 + c1) % 5, (c0 + c1) % 3. Here,
+(c0 + c1) is a common sub-expression that is evaluated only once. The output of
+printExprWithStats shows that the second instance is a duplicate. It is
+annotated with [CSE #2] which allows us to identify the original expression.
+Runtime stats for duplicate expressions are not displayed.
+
+.. code-block::
+
+    mod [cpu time: 50.49us, rows: 1024] -> BIGINT [#1]
+       cast(plus as BIGINT) [cpu time: 68.15us, rows: 1024] -> BIGINT [#2]
+          plus [cpu time: 51.84us, rows: 1024] -> INTEGER [#3]
+             c0 [cpu time: 0ns, rows: 0] -> INTEGER [#4]
+             c1 [cpu time: 0ns, rows: 0] -> INTEGER [#5]
+       5:BIGINT [cpu time: 0ns, rows: 0] -> BIGINT [#6]
+
+    mod [cpu time: 49.29us, rows: 1024] -> BIGINT [#7]
+       cast((plus(c0, c1)) as BIGINT) -> BIGINT [CSE #2]
+       3:BIGINT [cpu time: 0ns, rows: 0] -> BIGINT [#8]
+
+Another example is dictionary memoization. If we evaluate an expression over a
+column that uses the same base dictionary, the evaluation remembers the results
```

Comment:
Nit : Make explicit the expression evaluation happens over "multiple batches of data of the column using the same base dictionary". When reading I felt I got this context of multiple batches from the later sentences.

Reply:
Good suggestion. Updated.